File: generated-code/codecraft/csharp/TcpReadWrite/Model/AttackProperties.cs
namespace TransGenTest.Model
{
    /// <summary>
    /// Entity's attack properties
    /// </summary>
    public struct AttackProperties
    {
        /// <summary>
        /// Maximum attack range
        /// </summary>
        public int AttackRange { get; set; }
        /// <summary>
        /// Damage dealt in one tick
        /// </summary>
        public int Damage { get; set; }
        /// <summary>
        /// If true, dealing damage will collect resource from target
        /// </summary>
        public bool CollectResource { get; set; }
    
        public AttackProperties(int attackRange, int damage, bool collectResource)
        {
            this.AttackRange = attackRange;
            this.Damage = damage;
            this.CollectResource = collectResource;
        }
    
        /// <summary> Read AttackProperties from reader </summary>
        public static AttackProperties ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new AttackProperties();
            result.AttackRange = reader.ReadInt32();
            result.Damage = reader.ReadInt32();
            result.CollectResource = reader.ReadBoolean();
            return result;
        }
    
        /// <summary> Write AttackProperties to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(AttackRange);
            writer.Write(Damage);
            writer.Write(CollectResource);
        }
    
        /// <summary> Get string representation of AttackProperties </summary>
        public override string ToString() {
            string stringResult = "AttackProperties { ";
            stringResult += "AttackRange: ";
            stringResult += AttackRange.ToString();
            stringResult += ", ";
            stringResult += "Damage: ";
            stringResult += Damage.ToString();
            stringResult += ", ";
            stringResult += "CollectResource: ";
            stringResult += CollectResource.ToString();
            stringResult += " }";
            return stringResult;
        }
    }
}

File: generated-code/fsharp/codecraft/TcpReadWrite/Model/AttackProperties.fs
#nowarn "0058"

namespace TransGenTest.Model

open TransGenTest

/// Entity's attack properties
type AttackProperties = {
    /// Maximum attack range
    AttackRange: int;
    /// Damage dealt in one tick
    Damage: int;
    /// If true, dealing damage will collect resource from target
    CollectResource: bool;
} with

    /// Write AttackProperties to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.AttackRange
        writer.Write this.Damage
        writer.Write this.CollectResource

    /// Read AttackProperties from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        AttackRange = reader.ReadInt32()
        Damage = reader.ReadInt32()
        CollectResource = reader.ReadBoolean()
    }